RFC/src/RFC-0170_NetworkCommunicationProtocol.md

```diff
@@ -0,0 +1,301 @@
+
+Each CC or CN has a local routing table that contains the online communication addresses of all CCs and CNs on the Tari communication network known to that CC or CN.
+When a CC or CN wants to join the Tari communication network, the online communication address of at least one other CN that is part of the network needs to be known.
+The online communication address of the initial CN can either be manually provided or a bootstrapped list of "reliable" and persistent CNs can be provided with the Validator Node, Base Node, Wallet or Asset Manager software.
+The new CC or CN can then request additional peer contact information of other CNs from the initial peers to extend their own routing table.
+
+The routing table consists of a list of peer addresses that link node ids, public identification keys and online communication addresses of each known CC and CN.
+
+The Peer Address stored in the routing table can be implemented as follows:
+
+| Description      | Data type          | Comments                                                                 |
+|---               |---                 |---                                                                       |
+| network address  | network_address    | The online communication address of the CC or CN                         |
+| node_id          | node_id            | Registration Assigned for VN, Self selected for BN, W and AM             |
+| public_key       | public_key         | The public key of the identification cryptographic key of the CC or CN   |
+| node_type        | node_type          | VN, BN, W or AM                                                          |
+| linked asset ids | list of asset ids  | Asset ids can be used as an address on Tari network similar to a node id |
+| last_connection  | timestamp          | Time of last successful connection with peer                             |
+
+When a new CC or CN wants to join the Tari communication network they need to submit a joining request to the rest of the network.
+The joining request contains the peer address of the new CC or CN.
+Each CN that receives the joining request can decide if they want to add the new CCs or CNs contact information to their local routing table.
+When a CN, that received the joining request, has a similar node id to the new CC or CN then that node must add the peer address to their routing table.
+All CNs with similar node ids to the new CC or CN should have a copy of the new peer address in their routing tables.
+
+To limit potential attacks, only one registration with the same online communication address can be stored in the routing table of a CN.
+This restriction will limit Bad Actors from spinning up multiple CNs on a single computer.
```

Comment:
- A node id clash is very unlikely especially for node ids obtain from registration, derived node ids are a bit more open to manipulation. Generating a node id clash is as difficult as guessing someone's wallet address. Let's say I am a Bad Actor and want the same node id as another existing CC or CN, then I need to find a public key( and corresponding private key) that will generate the same node id. Having the private key for that public key is a must, otherwise you cant sign your joining request and it will be thrown out by the network. If someone validly generates a node id clash then the existing CC or CN is completely compromised, not just his network location.
- If a Bad CN does allow multiple registrations from the same online address in his local routing table it won't break the communication network. This is just a mechanism for the CN or CC to limit some unwanted registrations in their routing table, it could work without it.

Reply:
But what about when we have 2 node ID's registering on the same block? Are we including some unique number in there to allow this?

Reply:
I think we have two options: 
- Option 1: node ID=hash(hash(block)||public_key) => This allows more than one registration per block, but the Validator Node might have some control to manipulate its node ID. I think it will be difficult enough that it wont be an issue.

- Option 2: node_id=hash(block) => Limit the number of Validator Node registrations per block so that there is only one. All values used in node ID generation is out of the control of the new Validator Node.

Reply:
option 1 would be fine, you wont know the block hash before time